packages/shared/src/clients/rpc/RpcClient2.ts

```diff
@@ -0,0 +1,76 @@
+import { Logger } from '@l2beat/backend-tools'
+import { RateLimiter, UnixTime } from '@l2beat/shared-pure'
+import { generateId } from '../../tools/generateId'
+import { getBlockNumberAtOrBefore } from '../../tools/getBlockNumberAtOrBefore'
+import { HttpClient2 } from '../http/HttpClient2'
+import { Block, RpcResponse } from './types'
+
+interface RpcClient2Deps {
+  http: HttpClient2
+  logger: Logger
+  url: string
+  callsPerMinute?: number
+}
+
+export class RpcClient2 {
+  constructor(private readonly $: RpcClient2Deps) {
+    const rateLimiter = new RateLimiter({
+      callsPerMinute: this.$.callsPerMinute ?? 60,
+    })
+    this.query = rateLimiter.wrap(this.query.bind(this))
+    this.$.logger = this.$.logger.for(this)
+  }
+
+  // TODO: add support for including txs body
+  async getBlock(blockNumber: number | 'latest') {
+    const method = 'eth_getBlockByNumber'
+    const encodedNumber =
+      blockNumber === 'latest' ? 'latest' : `0x${blockNumber.toString(16)}`
+    const blockResponse = await this.query(method, [encodedNumber, false])
+
+    const block = Block.safeParse(blockResponse)
+    if (!block.success) {
+      this.$.logger.error(JSON.stringify(blockResponse))
+      throw new Error('Error during parsing of eth_getBlockByNumber response')
+    }
```

Comment:
This architecture of retries being implemented on HttpClient2 makes us unable to retry in this case. Also IIRC Etherscan is happy to return 200 OK { success: false } or something like that in their responses.

As an additional note it might be the case that some endpoints require more or less timeouts/retries and having a single http client underneath makes handling that impossible.

CC: @maciekop-l2b